### PR TITLE
Use the GenericJob interface by MPIJob

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This aligns the implementation of the MPIJob controller with Job controller. It also already drops some code that was duplicated.

#### Which issue(s) this PR fixes:

Part of #369

#### Special notes for your reviewer:

There will be follow ups:
- commonize the `Reconcile` functions
- commonize indexing
- commonize webhooks

### CODE REVIEW

**Comment by mimowo** [comment]
/assign @alculquicondor @kerthcet 
Please review, there will be follow ups, but it seems a sensible step on the path to the library for frameworks.

**Comment by mimowo** [comment]
cc @mwielgus

File: pkg/controller/workload/job/job_controller.go [hunk 1]
```diff
@@ -44,6 +44,7 @@ import (
 )
 
 var (
+	ownerKey          = ".metadata.controller"
```
Comment (alculquicondor) [inline pkg/controller/workload/job/job_controller.go:47]: this is an odd key. Generally it should match the jsonpath of where the value is found.
But in reality it doesn't matter.

Still, this should probably reference job in some way. What about `.metadata.ownerReferences[Job]`?

File: pkg/controller/workload/jobframework/helpers.go [hunk 2]
```diff
@@ -166,7 +166,11 @@ func StopJob(ctx context.Context, client client.Client, record record.EventRecor
 	}
 
 	if wl != nil {
-		if err := job.RestoreNodeAffinity([]map[string]string{wl.Spec.PodSets[0].Template.Spec.NodeSelector}); err != nil {
+		nodeSelectors := []map[string]string{}
+		for _, podSet := range wl.Spec.PodSets {
+			nodeSelectors = append(nodeSelectors, podSet.Template.Spec.NodeSelector)
+		}
+		if err := job.RestoreNodeAffinity(nodeSelectors); err != nil {
```
Comment (alculquicondor) [inline pkg/controller/workload/jobframework/helpers.go:173]: to avoid the allocation, I would just pass the list of PodSets to `RestoreNodeAffinity`

File: pkg/controller/workload/jobframework/interface.go [hunk 3]
```diff
@@ -59,4 +55,6 @@ type GenericJob interface {
 	PodsReady() bool
 	// GetWorkloadName returns the name of the workload for the job.
 	GetWorkloadName() string
+	// GetOwnerKey returns the name of the field indexing the workload's owner.
```
Comment (alculquicondor) [inline pkg/controller/workload/jobframework/interface.go:58]:
```suggestion
	// GetOwnerKey returns a unique key that identifies Workloads owned by the job implementation.
```

File: pkg/controller/workload/mpijob/mpijob_controller.go [hunk 4]
```diff
@@ -102,6 +98,177 @@ func NewReconciler(
...
+	job.Status.StartTime = nil
+	return true
+}
+
+func (job *MPIJob) GetOwnerKey() string {
+	return ownerKey
+}
+
+func (job *MPIJob) PodSets() []kueue.PodSet {
+	podSets := []kueue.PodSet{}
```
Comment (alculquicondor) [inline pkg/controller/workload/mpijob/mpijob_controller.go:157]: Use make to allocate the size

File: pkg/controller/workload/mpijob/mpijob_controller.go [hunk 5]
```diff
@@ -102,6 +98,177 @@ func NewReconciler(
...
+		if !equality.Semantic.DeepEqual(mpiReplicaSpec.Template.Spec.Containers,
+			wl.Spec.PodSets[index].Template.Spec.Containers) {
+			return false
+		}
+	}
+	return true
+}
+
+func (job *MPIJob) PriorityClass() string {
+	return calcPriorityClassName(&job.MPIJob)
```
Comment (alculquicondor) [inline pkg/controller/workload/mpijob/mpijob_controller.go:260]: any reason not to move the code here?

File: pkg/controller/workload/mpijob/mpijob_webhook.go [hunk 6]
```diff
@@ -112,7 +113,9 @@ func (w *MPIJobWebhook) ValidateUpdate(ctx context.Context, oldObj, newObj runti
 func validateUpdate(oldJob, newJob *kubeflow.MPIJob) field.ErrorList {
 	allErrs := validateCreate(newJob)
 
-	if !*newJob.Spec.RunPolicy.Suspend && (queueName(oldJob) != queueName(newJob)) {
+	oldMPIJob := MPIJob{*oldJob}
+	newMPIJob := MPIJob{*newJob}
```
Comment (alculquicondor) [inline pkg/controller/workload/mpijob/mpijob_webhook.go:117]: I still have the feeling that using an alias would save this kind of lines.

File: pkg/controller/workload/mpijob/mpijob_controller.go [hunk 7]
```diff
@@ -102,6 +98,177 @@ func NewReconciler(
...
+}
+
+func (job *MPIJob) PodsReady() bool {
```
Comment (kerthcet) [inline pkg/controller/workload/mpijob/mpijob_controller.go:263]: Not quite related to this PR, if MPIJob JobRunning is False, and we disable the `WaitForPodsReady`, will we still admit the workload?

Comment (mimowo) [reply pkg/controller/workload/mpijob/mpijob_controller.go:263]: I don't think I understand the question, can you maybe rephrase? Generally, `PodsReady` or `JobRunning` is `False` before admission, regardless of `WaitForPodsReady` being enabled.

File: pkg/controller/workload/job/job_controller.go (hunk 1 above)
Comment (mimowo) [reply pkg/controller/workload/job/job_controller.go]: Done, but I'm think that eventually, rather than requiring the frameworks to implement it based on GVK.

File: pkg/controller/workload/jobframework/interface.go (hunk 3 above)
Comment (mimowo) [reply pkg/controller/workload/jobframework/interface.go]: Done

File: pkg/controller/workload/mpijob/mpijob_controller.go (hunk 4 above)
Comment (mimowo) [reply pkg/controller/workload/mpijob/mpijob_controller.go]: Done

File: pkg/controller/workload/mpijob/mpijob_controller.go (hunk 5 above)
Comment (mimowo) [reply pkg/controller/workload/mpijob/mpijob_controller.go]: No, moved now.

File: pkg/controller/workload/mpijob/mpijob_webhook.go (hunk 6 above)
Comment (mimowo) [reply pkg/controller/workload/mpijob/mpijob_webhook.go:117]: I agree that methods like QueueName, or ParentWorkloadName belong to the Job indeed. However, methods for reconciling seem outside. Maybe we should have two types and split the methods between them. For example GenericJob and GenericJobReconciler. Then, I would be happy to use type aliases. I think this is something I would like to work as the next PR.

Comment (alculquicondor) [reply pkg/controller/workload/mpijob/mpijob_webhook.go:117]: sg

**Review by alculquicondor — COMMENTED** [review]
/lgtm
